File: 0876-hand-of-straights/0876-hand-of-straights.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        handCounter = defaultdict(int)
        for val in hand:
            handCounter[val] += 1
            
        allKeys =  list(handCounter.keys())
        heapq.heapify(allKeys)
        
        while allKeys:
            cur=allKeys[0]
            
            for i in range(cur,cur+groupSize):
                if i not in handCounter:
                    return False
                handCounter[i]-=1
                if handCounter[i]==0:
                    if i != allKeys[0]:
                        return False
                    heapq.heappop(allKeys)
                
            
        return True
                    
                
        
        

File: 0876-hand-of-straights/0876-hand-of-straights.cpp
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        int n = hand.size();
        map<int, int> hsh;

        for(auto val: hand){
            hsh[val]++;
        }

        if(n % groupSize != 0) return false;

        while(!hsh.empty()){
            int minVal = hsh.begin()->first;
            for(int i = 0; i < groupSize; i++){
                int currentVal = minVal + i;
                if(hsh.find(currentVal) == hsh.end()) return false;
                if(hsh[currentVal] == 1) hsh.erase(currentVal);
                else hsh[currentVal]--;
            }
        }
        return true;
    }
};